DjangoFilters: Add description for enum choices

valentijnscholten opened the pull request:
The choices in a ChoiceFilter (enum) can be documented via the description, similar to what's already in place for OrderingFilter.

This will help users, for example in SwaggerUI.

No idea if this is the best way / place to implement this, but it works locally at least.

Before:
![image](https://user-images.githubusercontent.com/4426050/119223579-8facf800-bafa-11eb-9eee-7c4b93101ede.png)


After:
![image](https://user-images.githubusercontent.com/4426050/119223594-9dfb1400-bafa-11eb-84c7-aca9d773f8d0.png)

tfranzel commented:
hi @valentijnscholten, that is a very nice idea and it has come up before in #337

as a matter of fact i do like this from a UI perspective, but i would prefer to have a structured approach via OpenAPI. others have raised this pain point in OAI/OpenAPI-Specification#681. maybe we should show some support there.

unfortunately it looks like it didn't make into 3.1.

valentijnscholten commented:
OK, didn't see that. I got this info from the official swagger docs, so I was expecting it to be common. Since it's just a few lines of code, I feel it wouldn't hurt to have this merged until (maybe in the distant future) there's a more structured solution from OA.

tfranzel commented on drf_spectacular/contrib/django_filters.py:
```diff
@@ -121,6 +121,12 @@ def resolve_filter_field(self, auto_schema, model, filterset_class, field_name,
 
+        if 'choices' in filter_field.extra:
+            description = description or ''
+            description += '>\n'
```
- the `>` looks weird.
- we should add an empty line as separation if there is an description
- i don't like the way the empty backticks look

valentijnscholten replied:
I thought the '>' and backticks where part of the official format needed, but it looks like they can be removed/changed, committed that. I did add the '\n' but it doesn't seem to do anything for the json spec, not even sure if json supports multiline strings". 

My first attempt was by modifying the `build_choice_field` in `plumbing.py`, but these fields from DjangoFilters don't seem to pass through there. Or do you suggest to add code there as well as keeping it in the django filter part?

tfranzel reviewed (changes requested):
it is true that this would would most likely not break anybody. 

we are still waiting for adoption of 3.1 in upstream projects like swagger-ui. i suppose 3.1.1 or 3.2 support is not gonna be available in the near future. with that said it's probably reasonable to support this in the meantime. it's just tricky to take features out again when the time comes, but the benefits probably outweigh the risk.

if we gonna move forward here, we would also need this for the non django-filter case. maybe factor out the the string generation into `plumbing.py` so we can use it in the other places too.

valentijnscholten commented:
I have extracted the description generator code and added a call in `publing.py#build_choice_field`. This seems to work for some fields, but not for all:

- A BaseCSVFilter based filter: still works OK as before: OK
- A ManyToManyRelatedField based filter: gets the description generated twice: NOK
- A model CharField with choides: response schema does not get the description: NOK

I am happy to help out further, but don't really have the time to fully dive into all the schema generation code in here. Do you maybe have some pointers of where this description would be best construct to fix these NOK cases @tfranzel ?

tfranzel commented:
hi @valentijnscholten, looks nice! please squash together the commits and force push for a clean history. i'll take over from there. can't really give pointers here. i would have to investigate myself on how to resolve the remaining issues. 

i'm pretty strapped for time at the moment. it will take a few days for me to get to it.

valentijnscholten commented:
sqaushed. thanks for looking into it, no rush, good to know it will eventually be implemented.

tfranzel commented:
superseded in favor of #952.

@valentijnscholten thank you for this PR. There were a couple more things to address with this feature, namely the postprocessing hook and non-django-filter behavior. Took me a while to get around to it. 

Please review this other PR whtether it adresses your points. Would be great if you could do it by end of the week, so that we can do a pending release.

tfranzel commented:
Ahh ok, sry you won't benefit from this anymore. Have a good one.

valentijnscholten commented:
Defect Dojo is still using it, so there will be benefit :-) Thanks for implementing the suggested enhancements.
